sockeye/utils.py

```diff
@@ -264,6 +264,148 @@ def average_tensors(tensors: List[pt.Tensor]) -> pt.Tensor:
     return sum(tensors) / len(tensors)  # type: ignore
 
 
+def gen_prefix_masking(prefix: pt.Tensor, vocab_size: int, dtype: pt.dtype) -> pt.Tensor:
+    """
+    Generate prefix masks from prefix ids, which are inf everywhere except zero for prefix ids.
+
+    :param prefix: Target prefix token or factors in ids. Shape (batch size, max length of prefix).
+    :param vocab_size: vocabulary size
+    :return prefix_masks (batch size, max length of prefix, vocab_size), with type as dtype
+
+    """
+    # create one hot vectors with size of vocab_size for each prefix id.
+    prefix_one_hot = pt.nn.functional.one_hot(prefix.to(pt.int64), num_classes=vocab_size).to(prefix.device)
+
+    # prefix_masks are inf everywhere except zero for indices of prefix ids.
+    prefix_masks = pt.full(prefix_one_hot.size(), fill_value=np.inf, device=prefix.device, dtype=dtype)
+    prefix_masks[prefix_one_hot != 0] = 0.
+
+    # In the same batch during inference, it is possible that some translations have target prefix
+    # while others do not have. It is also possible that translation may have a target prefix with
+    # different length to others. Thus prefix ids may include a full zero vector if a translation
+    # in the batch does not have prefix, or include a vector padding with zeros on the right if some
+    # translations are with shorter prefix. An example of prefix ids reflecting length differences \
+    # is as follows:
+    #
+    # [1, 2, 3]
+    # [1, 2, 0]
+    # [0, 0, 0]
+    #
+    # Here, the first sentence has a prefix of length 3, the second one has a prefix of length 1 \
+    # and the last one does not have prefix.
+    #
+    # At any timestep, some target prefix ids could be zero. If a prefix id is zero for a translation \
+    # at a timestep, all hots in the vocab are assigned to 0 (instead of only one hot is assigned to 0 \
+    # and other hots are inf). This makes sure there is no constraint on selecting any specific target \
+    # token for the translation in that case.
+
+    prefix_masks.masked_fill_(prefix.unsqueeze(-1) == 0, 0)
```

Comment:
Thanks Tobias. I think put 0 instead of PAD makes it more clear as 0 in the target_prefix explicitly means 'no constraint'. 

Take your suggestion into account I added more detail in the comment (before this line of code):

```
# At any timestep, some target prefix ids could be 0 (i.e. 0 in the target_prefix means 'no constraint'). \
# If a prefix id is 0 for a translation at a timestep, all hots in the vocab are assigned to 0 (instead \
# of only one hot is assigned to 0 and other hots are inf). This makes sure there is no constraint on \
# selecting any specific target token for the translation in that case.

prefix_masks.masked_fill_(prefix.unsqueeze(-1) == 0, 0)
```

Reply:
thanks!